Make alert widget not access all flash data (Fixes #250)

tobias-dietz opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #250

bizley commented on widgets/Alert.php:
```diff
@@ -51,21 +51,18 @@ class Alert extends \yii\bootstrap4\Widget
                     'options' => array_merge($this->options, [
-                        'id' => $this->getId() . '-' . $type . '-' . $i,
-                        'class' => $this->alertTypes[$type] . $appendClass,
+                        'id' => $this->getId().'-'.$type.'-'.$i,
```
Please keep the spaces, this is not Symfony 😀

tobias-dietz replied:
OK I'll remember that in future :smile: Do I have to push again in the branch I sent the pull request with the added spaces? (as said before, I'm new to GitHub and all the workflows and I don't want to mess up everything, sorry for that newbie questions :laughing: )

bizley replied:
Just push it to the same branch ;)

tobias-dietz commented:
Found the missing spaces and added them :smile: I've also added a unit test. The tests may seem verbose but it felt nessecary since the alert widget has a nested foreach loop and can process strings as well as arrays for the differnt message types. If it's too verbose, I would suggest to remove all tests except for the last three ones. 

The last test tests the fix for this bug and should fail with the current alert widget implementation in master but not with the fixed implementation.

bizley commented:
Wow, impressive, not everyday we can see so robust testing. One tiny thing - it should be PHPUnit tests strictly, not Codeception so please change the extended test case class and implementation details there (but it should be pretty straight forward if I'm not mistaken).
Also please add CHANGELOG line. Thank you!

tobias-dietz commented:
Thank you for your nice feedback :smile: I've pushed a new commit, the unit test looks now the same as the existing ones for the models (`tests\unit\models\LoginFormTest` etc.). 
However, I have the strong feeling that I did not understand exactly what you meant because those `expect()` are looking like they only wrap  Codeception methods. 

I tried to extend `PHPUnit\Framework\TestCase` but then I receive a lot of errors like `[yii\base\ErrorException] Attempt to read property "session" on null` (which wasn't that surprising to be honest).

Then I found https://github.com/yiisoft/yii2/blob/master/tests/TestCase.php which I think is what you meant? But it seems that this class is only part of `yiisoft/yii2` and not directly usable in `yiisoft/yii2-app-basic` (at least, PhpStorm doesn't offer that class on auto complete)? If I'm wrong (what is not that unlikely), it would be nice if you could point on additional resources or an example. I followed the TESTING section of the README file in `yiisoft/yii2-app-basic` so far.

Regarding the CHANGELOG: I couldn't find a CHANGELOG in `yiisoft/yii2-app-basic`.

bizley commented:
Oh, right, no changelog here, weird... Anyway, it looks like I went full-auto here and treated it just like Yii 2 core while it's not the case and the few tests here are indeed Codeception ones - could you please revert the changes I've told you to do and bring back the ones you made before? I'm very very sorry for that extra work 😭

tobias-dietz commented:
No problem :smile: The commit was broken either way (forgot to clean up my experiments with PHPUnit TestCase). Please have a look again. Basically, they are my original tests, only now with the Codeception `expect()` functions (instead of accessing `$this->tester`) that are also used in the other tests in yii2-app-basic (to maintain visual consistency with the other tests).

machour reviewed (approved):
Nice PR, thank you for your contribution.

samdark commented:
Thanks!

samdark commented:
Would you please apply the same to advanced app template?
